### PR TITLE
Fix NPU SDPA path silently discarding additive attention bias masks

### PR DESCRIPTION
**Pull request by junli-gong**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48843) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48843)

## What does this PR do?

On Ascend NPU, `sdpa_attention_forward` booleanizes any non-`torch.bool` `attention_mask` so the backend can dispatch to `FlashAttentionScore`:

```python
if _is_torch_npu_available:
    if attention_mask is not None and attention_mask.dtype != torch.bool:
        attention_mask = torch.logical_not(attention_mask.bool()).to(query.device)
```

This is correct for masks with **boolean semantics** (`0` = keep, `-inf` / `finfo.min` = mask out). It is destructive for masks that are an **additive attention bias**, where the magnitude of every entry carries information: `.bool()` maps every non-zero entry to `True`, so `logical_not(...)` yields an all-`False` tensor and the bias is silently dropped.

`ParakeetForCTC` hits this: `ParakeetAttention.forward` passes its relative-position score matrix as `attention_mask` (`modeling_parakeet.py` L338-L346), which is a documented, valid use of the float-mask form of `torch.nn.functional.scaled_dot_product_attention`. On NPU all 42 encoder layers therefore run with no positional bias at all.

## Evidence

`nvidia/parakeet-ctc-1.1b`, librispeech `test-other`, 200 samples, identical code and inputs on both sides. NPU = Ascend 910B3 (CANN 8.5.0, torch 2.6.0, torch_npu 2.6.0.post5); reference = NVIDIA A40.

| Setup | WER | edits | deletions |
|---|---|---|---|
| A40 (bf16) | 5.03% | 151 | 4 |
| NPU, current `main` | **15.12%** | 454 | **44** |
| NPU, with this patch | **4.96%** | 149 | **4** |

Forcing fp32 on NPU does not help (15.09%), which rules out a numerical-precision cause. With the patch, single-utterance decoding on NPU becomes token-identical to the A40 output.

Per-operator dumps (msprobe, same audio on both devices) localise it exactly: ops 0-106 match (cosine ~ 1), and the **first** SDPA call drops to cosine 0.42. Its `q`/`k`/`v` inputs are identical (cosine 0.99999); the only differing input is `attn_mask` - `bfloat16` in `[-37.0, +35.75]` on A40 vs `bool` all-`False` on NPU. The NPU trace also contains 42 extra `Torch.logical_not` calls (one per layer), the fingerprint of this branch.

## Fix

Apply the conversion only when the mask actually carries boolean semantics; leave additive bias masks untouched (they fall back to the small-operator path, i.e. correctness over speed).

## Tests

`tests/utils/test_sdpa_attention.py` patches the NPU availability flag, so the branch is exercised on any device (no NPU hardware needed):

- an additive bias mask must reach sdpa unchanged - fails on `main` with `torch.bool != torch.float32`, passes with this patch;
- `-inf` and `finfo.min` masks must still be booleanized, with `0 -> True` / masked -> `False` - passes both before and after, so existing behaviour is preserved.

## Trade-off / alternatives

The guard evaluates `.all()`, which forces one device sync per call on NPU. That seemed preferable to silently wrong outputs, but I'm happy to switch to either alternative if maintainers prefer:

1. an opt-out attribute on the attention module for models that pass additive biases;
2. migrating such models to the newer `position_bias` argument (handled after this branch by `create_position_bias_mask`, so it is unaffected today).

I don't have NPU CI access, so the numbers above are from a manual run on real hardware.

## Who can review?

@ArthurZucker @Rocketknight1 (attention / integrations)

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @IlyasMoutawwakil for Ascend!

File: tests/utils/test_sdpa_attention.py
Comment (IlyasMoutawwakil) [inline tests/utils/test_sdpa_attention.py]: we cant accept this file tho

File: src/transformers/integrations/sdpa_attention.py [hunk 1]
```diff
@@ -131,10 +131,23 @@ def sdpa_attention_forward(
...
+            if attention_mask.is_floating_point():
+                neg_fill = torch.finfo(attention_mask.dtype).min
+                has_boolean_semantics = (
+                    (attention_mask == 0) | (attention_mask == neg_fill) | attention_mask.isneginf()
+                ).all()
+            else:
+                has_boolean_semantics = True
+            if has_boolean_semantics:
+                # Convert to boolean type, making sdpa to force call FlashAttentionScore to improve performance.
+                attention_mask = torch.logical_not(attention_mask.bool()).to(query.device)
```
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/sdpa_attention.py]: sure as long as it's under npu

**Comment by IlyasMoutawwakil** [comment]
hi, do you work for ascend ? we dont have npu ascend to verify the claims

**Comment by junli-gong** [comment]
Hi @IlyasMoutawwakil, thanks for taking a look. No, I don't work for Huawei/Ascend. I ran into this while running Parakeet on Ascend 910B NPUs.

The part that decides correctness can be checked without an NPU. The NPU branch applies `torch.logical_not(attention_mask.bool())` to any non-bool mask, and for an additive bias every non-zero score ends up as `False`:

```python
>>> import torch
>>> bias = torch.tensor([[-37.0, 35.75], [1.5, -0.25]])  # additive relative-position scores
>>> torch.logical_not(bias.bool())
tensor([[False, False],
        [False, False]])
```

So the mask that reaches SDPA no longer depends on the bias values. The new tests patch `_is_torch_npu_available`, so they exercise this branch on CPU in the regular CI (they ran in `tests_non_model` on this PR). The additive-bias test fails on `main` and passes with this change. The two boolean-mask tests (`-inf` and `finfo.min`) pass on both, so those masks are still converted exactly as before.

What does need Ascend hardware is the end-to-end effect (the WER numbers in the description) and the cost of the extra `.all()` sync. @frozenleaves @FightingZhen, since this branch was added in #41143, would you be able to check it on an NPU? If there is a specific measurement you'd like from my side, I can run it on 910B and post the results.

**Comment by frozenleaves** [comment]
> Hi @IlyasMoutawwakil, thanks for taking a look. No, I don't work for Huawei/Ascend. I ran into this while running Parakeet on Ascend 910B NPUs.
> 
> The part that decides correctness can be checked without an NPU. The NPU branch applies `torch.logical_not(attention_mask.bool())` to any non-bool mask, and for an additive bias every non-zero score ends up as `False`:
> 
> ```python
> >>> import torch
> >>> bias = torch.tensor([[-37.0, 35.75], [1.5, -0.25]])  # additive relative-position scores
> >>> torch.logical_not(bias.bool())
> tensor([[False, False],
>         [False, False]])
> ```
> 
> So the mask that reaches SDPA no longer depends on the bias values. The new tests patch `_is_torch_npu_available`, so they exercise this branch on CPU in the regular CI (they ran in `tests_non_model` on this PR). The additive-bias test fails on `main` and passes with this change. The two boolean-mask tests (`-inf` and `finfo.min`) pass on both, so those masks are still converted exactly as before.
> 
> What does need Ascend hardware is the end-to-end effect (the WER numbers in the description) and the cost of the extra `.all()` sync. @frozenleaves @FightingZhen, since this branch was added in #41143, would you be able to check it on an NPU? If there is a specific measurement you'd like from my side, I can run it on 910B and post the results.

This PR was originally intended to integrate the SDPA interface into the FA backend, which may indeed have an impact on the additive attention bias. Since this PR was submitted quite early, if you have identified the root cause of the issue, it is necessary to fix it. Thank you for pointing this out. However, please note that when making the fix, remember to test the end-to-end training accuracy, and you can force the SDPA interface to call the FlashAttentionScore operator for comparison. Thanks for your hard work.

**Comment by junli-gong** [comment]
Thanks @frozenleaves! Just want to make sure I run the right comparison. This change only affects additive bias masks like the one Parakeet uses, while regular padding and causal masks behave exactly as before. For Parakeet, forcing FlashAttentionScore basically means turning the bias into a boolean mask, which is what main does today. So I'm planning a short fine-tuning run of the 1.1B Parakeet CTC model on 910B, once on main and once with this PR, using the same data, seed and number of steps. I'll compare the loss curves and WER against a GPU run and also report the throughput of both NPU runs. Does that match what you had in mind?
